// html2canvas-pro/dist/lib/render/canvas/text/text-decoration-renderer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TextDecorationRenderer = void 0;
const color_utilities_1 = require("../../../css/types/color-utilities");
class TextDecorationRenderer {
    constructor(ctx) {
        this.ctx = ctx;
    }
    render(bounds, styles) {
        this.ctx.fillStyle = (0, color_utilities_1.asString)(styles.textDecorationColor || styles.color);
        let thickness = 1;
        if (typeof styles.textDecorationThickness === 'number') {
            thickness = styles.textDecorationThickness;
        }
        else if (styles.textDecorationThickness === 'from-font') {
            thickness = Math.max(1, Math.floor(styles.fontSize.number * 0.05));
        }
        let underlineOffset = 0;
        if (typeof styles.textUnderlineOffset === 'number') {
            underlineOffset = styles.textUnderlineOffset;
        }
        const decorationStyle = styles.textDecorationStyle;
        styles.textDecorationLine.forEach((line) => {
            let y = 0;
            switch (line) {
                case 1 /* TEXT_DECORATION_LINE.UNDERLINE */:
                    y = bounds.top + bounds.height - thickness + underlineOffset;
                    break;
                case 2 /* TEXT_DECORATION_LINE.OVERLINE */:
                    y = bounds.top;
                    break;
                case 3 /* TEXT_DECORATION_LINE.LINE_THROUGH */:
                    y = bounds.top + (bounds.height / 2 - thickness / 2);
                    break;
                default:
                    return;
            }
            this.draw(bounds.left, y, bounds.width, thickness, decorationStyle);
        });
    }
    draw(x, y, width, thickness, style) {
        switch (style) {
            case 0 /* TEXT_DECORATION_STYLE.SOLID */:
                this.ctx.fillRect(x, y, width, thickness);
                break;
            case 1 /* TEXT_DECORATION_STYLE.DOUBLE */: {
                const gap = Math.max(1, thickness);
                this.ctx.fillRect(x, y, width, thickness);
                this.ctx.fillRect(x, y + thickness + gap, width, thickness);
                break;
            }
            case 2 /* TEXT_DECORATION_STYLE.DOTTED */:
                this.drawPattern(x, y, width, thickness, [thickness, thickness * 2]);
                break;
            case 3 /* TEXT_DECORATION_STYLE.DASHED */:
                this.drawPattern(x, y, width, thickness, [thickness * 3, thickness * 2]);
                break;
            case 4 /* TEXT_DECORATION_STYLE.WAVY */:
                this.drawWavy(x, y, width, thickness);
                break;
            default:
                this.ctx.fillRect(x, y, width, thickness);
        }
    }
    drawPattern(x, y, width, thickness, dash) {
        this.ctx.save();
        this.ctx.beginPath();
        this.ctx.setLineDash(dash);
        this.ctx.lineWidth = thickness;
        this.ctx.strokeStyle = this.ctx.fillStyle;
        this.ctx.moveTo(x, y + thickness / 2);
        this.ctx.lineTo(x + width, y + thickness / 2);
        this.ctx.stroke();
        this.ctx.restore();
    }
    drawWavy(x, y, width, thickness) {
        this.ctx.save();
        this.ctx.beginPath();
        this.ctx.lineWidth = thickness;
        this.ctx.strokeStyle = this.ctx.fillStyle;
        const amplitude = thickness * 2;
        const wavelength = thickness * 4;
        let currentX = x;
        this.ctx.moveTo(currentX, y + thickness / 2);
        while (currentX < x + width) {
            const nextX = Math.min(currentX + wavelength / 2, x + width);
            this.ctx.quadraticCurveTo(currentX + wavelength / 4, y + thickness / 2 - amplitude, nextX, y + thickness / 2);
            currentX = nextX;
            if (currentX < x + width) {
                const nextX2 = Math.min(currentX + wavelength / 2, x + width);
                this.ctx.quadraticCurveTo(currentX + wavelength / 4, y + thickness / 2 + amplitude, nextX2, y + thickness / 2);
                currentX = nextX2;
            }
        }
        this.ctx.stroke();
        this.ctx.restore();
    }
}
exports.TextDecorationRenderer = TextDecorationRenderer;
