// html2canvas-pro/dist/lib/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Html2CanvasConfig = void 0;
/**
 * Html2Canvas Configuration
 *
 * Manages configuration state for html2canvas rendering.
 * Eliminates the need for global static variables.
 */
class Html2CanvasConfig {
    constructor(options = {}) {
        const resolvedWindow = options.window || (typeof window !== 'undefined' ? window : undefined);
        if (!resolvedWindow) {
            throw new Error('Window object is required but not available');
        }
        this.window = resolvedWindow;
        this.cspNonce = options.cspNonce;
        this.cache = options.cache;
    }
    /**
     * Create configuration from an element
     * Extracts window from element's owner document
     */
    static fromElement(element, options = {}) {
        const ownerDocument = element.ownerDocument;
        if (!ownerDocument) {
            throw new Error('Element is not attached to a document');
        }
        const defaultView = ownerDocument.defaultView;
        if (!defaultView) {
            throw new Error('Document is not attached to a window');
        }
        return new Html2CanvasConfig({
            window: defaultView,
            ...options
        });
    }
    /**
     * Clone configuration with override options
     */
    clone(options = {}) {
        return new Html2CanvasConfig({
            window: options.window || this.window,
            cspNonce: options.cspNonce ?? this.cspNonce,
            cache: options.cache ?? this.cache
        });
    }
}
exports.Html2CanvasConfig = Html2CanvasConfig;
