// html2canvas-pro/dist/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PerformanceMonitor = exports.createDefaultValidator = exports.Validator = exports.Html2CanvasConfig = exports.html2canvas = void 0;
const config_1 = require("./config");
Object.defineProperty(exports, "Html2CanvasConfig", { enumerable: true, get: function () { return config_1.Html2CanvasConfig; } });
const validator_1 = require("./core/validator");
Object.defineProperty(exports, "createDefaultValidator", { enumerable: true, get: function () { return validator_1.createDefaultValidator; } });
Object.defineProperty(exports, "Validator", { enumerable: true, get: function () { return validator_1.Validator; } });
const performance_monitor_1 = require("./core/performance-monitor");
Object.defineProperty(exports, "PerformanceMonitor", { enumerable: true, get: function () { return performance_monitor_1.PerformanceMonitor; } });
const render_element_1 = require("./core/render-element");
/**
 * Renders an HTML element to a `<canvas>` element.
 *
 * The function clones the target element and its subtree into a hidden iframe,
 * resolves all computed styles, and paints the result onto a canvas —
 * producing a visual snapshot of the DOM as it appears in the browser.
 *
 * @example
 * ```ts
 * import html2canvas from 'html2canvas-pro';
 *
 * const canvas = await html2canvas(document.body, {
 *   backgroundColor: '#ffffff',
 *   scale: 2,
 *   useCORS: true
 * });
 * document.body.appendChild(canvas);
 * ```
 *
 * @param element - The root HTMLElement to render.
 * @param options - Rendering options.
 * @param config  - Advanced configuration. In most cases this is auto-created;
 *                  only pass it if you need to share a cache across multiple calls.
 * @returns A promise that resolves to the rendered HTMLCanvasElement.
 *
 * @throws {Error} If the element is not attached to a document.
 * @throws {DOMException} If an {@link Options.signal|AbortSignal} was provided and the operation was aborted.
 */
const html2canvas = (element, options = {}, config) => {
    const finalConfig = config ||
        config_1.Html2CanvasConfig.fromElement(element, {
            cspNonce: options.cspNonce,
            cache: options.cache
        });
    return (0, render_element_1.renderElement)(element, options, finalConfig);
};
exports.html2canvas = html2canvas;
exports.default = html2canvas;
