// html2canvas-pro/dist/lib/core/cache-storage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Cache = void 0;
const features_1 = require("./features");
class Cache {
    constructor(context, _options) {
        this.context = context;
        this._options = _options;
        this._cache = new Map();
        this._pendingOperations = new Map();
        // Default cache size: 100 items
        this.maxSize = _options.maxCacheSize ?? 100;
        if (this.maxSize < 1) {
            throw new Error('Cache maxSize must be at least 1');
        }
        if (this.maxSize > 10000) {
            this.context.logger.warn(`Cache maxSize ${this.maxSize} is very large and may cause memory issues. ` +
                `Consider using a smaller value (recommended: 100-1000).`);
        }
    }
    addImage(src) {
        // Wait for any pending operations on this key
        const pending = this._pendingOperations.get(src);
        if (pending) {
            return pending;
        }
        if (this.has(src)) {
            // Move to end for LRU ordering
            const entry = this._cache.get(src);
            this._cache.delete(src);
            this._cache.set(src, entry);
            return Promise.resolve();
        }
        if (isBlobImage(src) || isRenderable(src)) {
            // Create a pending operation to ensure atomicity
            const operation = this._addImageInternal(src);
            this._pendingOperations.set(src, operation);
            operation.finally(() => {
                this._pendingOperations.delete(src);
            });
            return operation;
        }
        return Promise.resolve();
    }
    async _addImageInternal(src) {
        const timeoutMs = this._options.imageTimeout ?? 15000;
        const imageWithTimeout = this.withTimeout(this.loadImage(src), timeoutMs, `Timed out (${timeoutMs}ms) loading image`);
        // Handle errors to prevent unhandled rejections
        imageWithTimeout.catch((error) => {
            this.context.logger.error(`Failed to load image ${src}: ${error instanceof Error ? error.message : 'Unknown error'}`);
        });
        // Store the promise with timeout in cache
        this.set(src, imageWithTimeout);
    }
    withTimeout(promise, timeoutMs, message) {
        if (timeoutMs <= 0) {
            return promise;
        }
        let timeoutId;
        const timeout = new Promise((_, reject) => {
            timeoutId = setTimeout(() => reject(new Error(message)), timeoutMs);
        });
        return Promise.race([promise, timeout]).finally(() => {
            if (timeoutId !== undefined) {
                clearTimeout(timeoutId);
            }
        });
    }
    match(src) {
        const entry = this._cache.get(src);
        if (entry) {
            // Move to end for LRU ordering (O(1))
            this._cache.delete(src);
            this._cache.set(src, entry);
            return entry.value;
        }
        return undefined;
    }
    /**
     * Set a value in cache with LRU eviction (O(1) via Map insertion order).
     * Map preserves insertion order; delete+set on access moves items to the end.
     * The first key in Map.keys() is always the least recently used.
     */
    set(key, value) {
        if (this._cache.has(key)) {
            // Update existing entry: move to end of Map
            this._cache.delete(key);
        }
        else if (this._cache.size >= this.maxSize) {
            // Evict LRU (first key = least recently used) — O(1)
            const lruKey = this._cache.keys().next().value;
            if (lruKey !== undefined) {
                this._cache.delete(lruKey);
                this.context.logger.debug(`Cache: Evicted LRU entry: ${lruKey}`);
            }
        }
        this._cache.set(key, { value });
    }
    /**
     * Get cache size
     */
    size() {
        return this._cache.size;
    }
    /**
     * Get max cache size
     */
    getMaxSize() {
        return this.maxSize;
    }
    /**
     * Clear all cache entries
     */
    clear() {
        this._cache.clear();
    }
    async loadImage(key) {
        const originChecker = this.context.originChecker;
        const defaultIsSameOrigin = (src) => originChecker.isSameOrigin(src);
        const isSameOrigin = typeof this._options.customIsSameOrigin === 'function'
            ? await this._options.customIsSameOrigin(key, defaultIsSameOrigin)
            : defaultIsSameOrigin(key);
        const useCORS = !isInlineImage(key) && this._options.useCORS === true && features_1.FEATURES.SUPPORT_CORS_IMAGES && !isSameOrigin;
        const useProxy = !isInlineImage(key) &&
            !isSameOrigin &&
            !isBlobImage(key) &&
            typeof this._options.proxy === 'string' &&
            features_1.FEATURES.SUPPORT_CORS_XHR &&
            !useCORS;
        if (!isSameOrigin &&
            this._options.allowTaint === false &&
            !isInlineImage(key) &&
            !isBlobImage(key) &&
            !useProxy &&
            !useCORS) {
            return;
        }
        let src = key;
        if (useProxy) {
            src = await this.proxy(src);
        }
        this.context.logger.debug(`Added image ${key.substring(0, 256)}`);
        return await new Promise((resolve, reject) => {
            const img = new Image();
            img.onload = () => resolve(img);
            img.onerror = reject;
            //ios safari 10.3 taints canvas with data urls unless crossOrigin is set to anonymous
            if (isInlineBase64Image(src) || useCORS) {
                img.crossOrigin = 'anonymous';
            }
            img.src = src;
            if (img.complete === true) {
                // Inline XML images may fail to parse, throwing an Error later on
                setTimeout(() => resolve(img), 500);
            }
        });
    }
    has(key) {
        return this._cache.has(key);
    }
    keys() {
        return Promise.resolve(Array.from(this._cache.keys()));
    }
    proxy(src) {
        const proxy = this._options.proxy;
        if (!proxy) {
            throw new Error('No proxy defined');
        }
        const key = src.substring(0, 256);
        return new Promise((resolve, reject) => {
            const responseType = features_1.FEATURES.SUPPORT_RESPONSE_TYPE ? 'blob' : 'text';
            const xhr = new XMLHttpRequest();
            xhr.onload = () => {
                if (xhr.status === 200) {
                    if (responseType === 'text') {
                        resolve(xhr.response);
                    }
                    else {
                        const reader = new FileReader();
                        reader.addEventListener('load', () => resolve(reader.result), false);
                        reader.addEventListener('error', (e) => reject(e), false);
                        reader.readAsDataURL(xhr.response);
                    }
                }
                else {
                    reject(`Failed to proxy resource ${key} with status code ${xhr.status}`);
                }
            };
            xhr.onerror = reject;
            const queryString = proxy.indexOf('?') > -1 ? '&' : '?';
            xhr.open('GET', `${proxy}${queryString}url=${encodeURIComponent(src)}&responseType=${responseType}`);
            if (responseType !== 'text' && xhr instanceof XMLHttpRequest) {
                xhr.responseType = responseType;
            }
            if (this._options.imageTimeout) {
                const timeout = this._options.imageTimeout;
                xhr.timeout = timeout;
                xhr.ontimeout = () => reject(`Timed out (${timeout}ms) proxying ${key}`);
            }
            xhr.send();
        });
    }
}
exports.Cache = Cache;
const INLINE_SVG = /^data:image\/svg\+xml/i;
const INLINE_BASE64 = /^data:image\/.*;base64,/i;
const INLINE_IMG = /^data:image\/.*/i;
const isRenderable = (src) => features_1.FEATURES.SUPPORT_SVG_DRAWING || !isSVG(src);
const isInlineImage = (src) => INLINE_IMG.test(src);
const isInlineBase64Image = (src) => INLINE_BASE64.test(src);
const isBlobImage = (src) => src.substring(0, 4) === 'blob';
const isSVG = (src) => src.slice(-3).toLowerCase() === 'svg' || INLINE_SVG.test(src);
