// html2canvas-pro/dist/lib/core/constants.js
"use strict";
/**
 * Shared numeric constants used across html2canvas-pro.
 * Centralising magic numbers for maintainability.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.RESOURCE_KEY_LOG_LENGTH = exports.INLINE_IMAGE_RESOLVE_DELAY_MS = exports.IFRAME_READY_POLL_MS = exports.SHADOW_MASK_OFFSET = exports.DEFAULT_IMAGE_TIMEOUT_MS = exports.MAX_IMAGE_CACHE_SIZE = exports.DEFAULT_IMAGE_CACHE_SIZE = exports.PATTERN_CACHE_MAX = exports.PARSE_CACHE_MAX_PER_DESCRIPTOR = void 0;
/** Maximum entries in the CSS parse cache per descriptor (LRU). */
exports.PARSE_CACHE_MAX_PER_DESCRIPTOR = 200;
/** Maximum cached background-image patterns per render session (LRU). */
exports.PATTERN_CACHE_MAX = 50;
/** Maximum size of the image-resource cache (entries). */
exports.DEFAULT_IMAGE_CACHE_SIZE = 100;
/** Maximum allowed image-resource cache size. */
exports.MAX_IMAGE_CACHE_SIZE = 10000;
/** Default image-load timeout in milliseconds. */
exports.DEFAULT_IMAGE_TIMEOUT_MS = 15000;
/** Z-axis mask offset used during box-shadow rendering. */
exports.SHADOW_MASK_OFFSET = 10000;
/** Polling interval (ms) when waiting for the cloned iframe to become ready. */
exports.IFRAME_READY_POLL_MS = 50;
/** Deferred resolution delay (ms) for inline XML images that may fail to parse. */
exports.INLINE_IMAGE_RESOLVE_DELAY_MS = 500;
/** Maximum characters logged for image/resource keys. */
exports.RESOURCE_KEY_LOG_LENGTH = 256;
