// html2canvas-pro/dist/lib/render/canvas/content-renderer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.renderReplacedElements = renderReplacedElements;
exports.renderFormElements = renderFormElements;
exports.renderListMarker = renderListMarker;
const color_utilities_1 = require("../../css/types/color-utilities");
const image_element_container_1 = require("../../dom/replaced-elements/image-element-container");
const canvas_element_container_1 = require("../../dom/replaced-elements/canvas-element-container");
const svg_element_container_1 = require("../../dom/replaced-elements/svg-element-container");
const iframe_element_container_1 = require("../../dom/replaced-elements/iframe-element-container");
const input_element_container_1 = require("../../dom/replaced-elements/input-element-container");
const textarea_element_container_1 = require("../../dom/elements/textarea-element-container");
const select_element_container_1 = require("../../dom/elements/select-element-container");
const bounds_1 = require("../../css/layout/bounds");
const text_1 = require("../../css/layout/text");
const vector_1 = require("../vector");
const box_sizing_1 = require("../box-sizing");
const image_rendering_1 = require("../../css/property-descriptors/image-rendering");
const length_percentage_1 = require("../../css/types/length-percentage");
const line_height_1 = require("../../css/property-descriptors/line-height");
const bitwise_1 = require("../../core/bitwise");
/**
 * Render replaced elements: Image, Canvas, SVG, IFrame.
 */
async function renderReplacedElements(ctx, context, options, iframeRendererFactory, container, curves, styles, renderReplacedElementFn) {
    if (container instanceof image_element_container_1.ImageElementContainer) {
        try {
            const image = await context.cache.match(container.src);
            const prevSmoothing = ctx.imageSmoothingEnabled;
            if (styles.imageRendering === image_rendering_1.IMAGE_RENDERING.PIXELATED ||
                styles.imageRendering === image_rendering_1.IMAGE_RENDERING.CRISP_EDGES) {
                ctx.imageSmoothingEnabled = false;
            }
            else if (styles.imageRendering === image_rendering_1.IMAGE_RENDERING.SMOOTH) {
                ctx.imageSmoothingEnabled = true;
            }
            renderReplacedElementFn(container, curves, image);
            ctx.imageSmoothingEnabled = prevSmoothing;
        }
        catch (e) {
            context.logger.error(`Error loading image ${container.src}`);
            context.onError?.(e instanceof Error ? e : new Error(String(e)));
        }
    }
    if (container instanceof canvas_element_container_1.CanvasElementContainer) {
        renderReplacedElementFn(container, curves, container.canvas);
    }
    if (container instanceof svg_element_container_1.SVGElementContainer) {
        try {
            const image = await context.cache.match(container.svg);
            renderReplacedElementFn(container, curves, image);
        }
        catch (e) {
            context.logger.error(`Error loading svg ${container.svg.substring(0, 255)}`);
            context.onError?.(e instanceof Error ? e : new Error(String(e)));
        }
    }
    if (container instanceof iframe_element_container_1.IFrameElementContainer && container.tree) {
        const iframeRenderer = iframeRendererFactory(context, {
            scale: options.scale,
            backgroundColor: container.backgroundColor,
            x: 0,
            y: 0,
            width: container.width,
            height: container.height
        });
        const canvas = await iframeRenderer.render(container.tree);
        if (container.width && container.height) {
            ctx.drawImage(canvas, 0, 0, container.width, container.height, container.bounds.left, container.bounds.top, container.bounds.width, container.bounds.height);
        }
    }
}
/**
 * Render form element content: checkbox, radio, text input.
 */
function renderFormElements(ctx, fontMetrics, textRenderer, pathFn, container, styles) {
    if (container instanceof input_element_container_1.InputElementContainer) {
        const size = Math.min(container.bounds.width, container.bounds.height);
        if (container.type === input_element_container_1.CHECKBOX && container.checked) {
            ctx.save();
            pathFn([
                new vector_1.Vector(container.bounds.left + size * 0.39363, container.bounds.top + size * 0.79),
                new vector_1.Vector(container.bounds.left + size * 0.16, container.bounds.top + size * 0.5549),
                new vector_1.Vector(container.bounds.left + size * 0.27347, container.bounds.top + size * 0.44071),
                new vector_1.Vector(container.bounds.left + size * 0.39694, container.bounds.top + size * 0.5649),
                new vector_1.Vector(container.bounds.left + size * 0.72983, container.bounds.top + size * 0.23),
                new vector_1.Vector(container.bounds.left + size * 0.84, container.bounds.top + size * 0.34085),
                new vector_1.Vector(container.bounds.left + size * 0.39363, container.bounds.top + size * 0.79)
            ]);
            ctx.fillStyle = (0, color_utilities_1.asString)(input_element_container_1.INPUT_COLOR);
            ctx.fill();
            ctx.restore();
        }
        else if (container.type === input_element_container_1.RADIO && container.checked) {
            ctx.save();
            ctx.beginPath();
            ctx.arc(container.bounds.left + size / 2, container.bounds.top + size / 2, size / 4, 0, Math.PI * 2, true);
            ctx.fillStyle = (0, color_utilities_1.asString)(input_element_container_1.INPUT_COLOR);
            ctx.fill();
            ctx.restore();
        }
    }
    if (isTextInputElement(container) && container.value.length) {
        const [font, fontFamily, fontSize] = textRenderer.createFontStyle(styles);
        const { baseline } = fontMetrics.getMetrics(fontFamily, fontSize);
        ctx.font = font;
        const isPlaceholder = container instanceof input_element_container_1.InputElementContainer && container.isPlaceholder;
        ctx.fillStyle = isPlaceholder ? (0, color_utilities_1.asString)(input_element_container_1.PLACEHOLDER_COLOR) : (0, color_utilities_1.asString)(styles.color);
        ctx.textBaseline = 'alphabetic';
        ctx.textAlign = canvasTextAlign(container.styles.textAlign);
        const bounds = (0, box_sizing_1.contentBox)(container);
        let x = 0;
        switch (container.styles.textAlign) {
            case 1 /* TEXT_ALIGN.CENTER */:
                x += bounds.width / 2;
                break;
            case 2 /* TEXT_ALIGN.RIGHT */:
                x += bounds.width;
                break;
        }
        let verticalOffset = 0;
        if (container instanceof input_element_container_1.InputElementContainer) {
            const fontSizeValue = (0, length_percentage_1.getAbsoluteValue)(styles.fontSize, 0);
            verticalOffset = (bounds.height - fontSizeValue) / 2;
        }
        const textBounds = bounds.add(x, verticalOffset, 0, 0);
        ctx.save();
        pathFn([
            new vector_1.Vector(bounds.left, bounds.top),
            new vector_1.Vector(bounds.left + bounds.width, bounds.top),
            new vector_1.Vector(bounds.left + bounds.width, bounds.top + bounds.height),
            new vector_1.Vector(bounds.left, bounds.top + bounds.height)
        ]);
        ctx.clip();
        textRenderer.renderTextWithLetterSpacing(new text_1.TextBounds(container.value, textBounds), styles.letterSpacing, baseline, styles.writingMode);
        ctx.restore();
        ctx.textBaseline = 'alphabetic';
        ctx.textAlign = 'left';
    }
}
/**
 * Render list-item marker (image or text).
 */
async function renderListMarker(ctx, context, textRenderer, paint, container, styles) {
    if (!(0, bitwise_1.contains)(container.styles.display, 2048 /* DISPLAY.LIST_ITEM */)) {
        return;
    }
    if (container.styles.listStyleImage !== null) {
        const img = container.styles.listStyleImage;
        if (img.type === 0 /* CSSImageType.URL */) {
            const url = img.url;
            try {
                const image = await context.cache.match(url);
                ctx.drawImage(image, container.bounds.left - (image.width + 10), container.bounds.top);
            }
            catch (e) {
                context.logger.error(`Error loading list-style-image ${url}`);
                context.onError?.(e instanceof Error ? e : new Error(String(e)));
            }
        }
    }
    else if (paint.listValue && container.styles.listStyleType !== -1 /* LIST_STYLE_TYPE.NONE */) {
        const [font] = textRenderer.createFontStyle(styles);
        ctx.font = font;
        ctx.fillStyle = (0, color_utilities_1.asString)(styles.color);
        ctx.textBaseline = 'middle';
        ctx.textAlign = 'right';
        const bounds = new bounds_1.Bounds(container.bounds.left, container.bounds.top + (0, length_percentage_1.getAbsoluteValue)(container.styles.paddingTop, container.bounds.width), container.bounds.width, (0, line_height_1.computeLineHeight)(styles.lineHeight, styles.fontSize.number) / 2 + 1);
        textRenderer.renderTextWithLetterSpacing(new text_1.TextBounds(paint.listValue, bounds), styles.letterSpacing, (0, line_height_1.computeLineHeight)(styles.lineHeight, styles.fontSize.number) / 2 + 2, styles.writingMode);
        ctx.textBaseline = 'bottom';
        ctx.textAlign = 'left';
    }
}
/**
 * Type guard for text input containers.
 */
const isTextInputElement = (container) => {
    if (container instanceof textarea_element_container_1.TextareaElementContainer) {
        return true;
    }
    else if (container instanceof select_element_container_1.SelectElementContainer) {
        return true;
    }
    else if (container instanceof input_element_container_1.InputElementContainer && container.type !== input_element_container_1.RADIO && container.type !== input_element_container_1.CHECKBOX) {
        return true;
    }
    return false;
};
/**
 * Map CSS text-align to Canvas textAlign.
 */
const canvasTextAlign = (textAlign) => {
    switch (textAlign) {
        case 1 /* TEXT_ALIGN.CENTER */:
            return 'center';
        case 2 /* TEXT_ALIGN.RIGHT */:
            return 'right';
        case 0 /* TEXT_ALIGN.LEFT */:
        default:
            return 'left';
    }
};
