// html2canvas-pro/dist/lib/css/property-descriptors/background-blend-mode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.backgroundBlendMode = exports.MIX_BLEND_MODE = void 0;
const parser_1 = require("../syntax/parser");
exports.MIX_BLEND_MODE = {
    NORMAL: 'normal',
    MULTIPLY: 'multiply',
    SCREEN: 'screen',
    OVERLAY: 'overlay',
    DARKEN: 'darken',
    LIGHTEN: 'lighten',
    COLOR_DODGE: 'color-dodge',
    COLOR_BURN: 'color-burn',
    HARD_LIGHT: 'hard-light',
    SOFT_LIGHT: 'soft-light',
    DIFFERENCE: 'difference',
    EXCLUSION: 'exclusion',
    HUE: 'hue',
    SATURATION: 'saturation',
    COLOR: 'color',
    LUMINOSITY: 'luminosity'
};
const VALID_VALUES = new Set(Object.values(exports.MIX_BLEND_MODE));
exports.backgroundBlendMode = {
    name: 'background-blend-mode',
    initialValue: 'normal',
    prefix: false,
    type: 1 /* PropertyDescriptorParsingType.LIST */,
    parse: (_context, tokens) => {
        const modes = [];
        const filtered = tokens.filter(parser_1.nonFunctionArgSeparator);
        for (const token of filtered) {
            if ((0, parser_1.isIdentToken)(token)) {
                if (VALID_VALUES.has(token.value)) {
                    modes.push(token.value);
                }
            }
        }
        return modes.length > 0 ? modes : ['normal'];
    }
};
