// html2canvas-pro/dist/lib/css/property-descriptors/border-image-slice.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.borderImageSlice = void 0;
const parser_1 = require("../syntax/parser");
const fillSides = (sides) => {
    const result = [...sides];
    if (result.length === 1) {
        result.push(result[0], result[0], result[0]);
    }
    else if (result.length === 2) {
        result.push(result[0], result[1]);
    }
    else if (result.length === 3) {
        result.push(result[1]);
    }
    return result.slice(0, 4);
};
exports.borderImageSlice = {
    name: 'border-image-slice',
    initialValue: '100%',
    prefix: false,
    type: 1 /* PropertyDescriptorParsingType.LIST */,
    parse: (_context, tokens) => {
        const filtered = tokens.filter(parser_1.nonWhiteSpace);
        const values = [];
        let fill = false;
        let unit = 'percent';
        for (const token of filtered) {
            if ((0, parser_1.isIdentToken)(token) && token.value === 'fill') {
                fill = true;
                continue;
            }
            if (token.type === 17 /* TokenType.NUMBER_TOKEN */) {
                values.push(token.number);
                unit = 'number';
            }
            else if (token.type === 16 /* TokenType.PERCENTAGE_TOKEN */) {
                values.push(token.number);
                unit = 'percent';
            }
        }
        const [top, right, bottom, left] = fillSides(values.length > 0 ? values : [100]);
        return { top, right, bottom, left, fill, unit };
    }
};
