// html2canvas-pro/dist/lib/css/property-descriptors/border-image-repeat.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.borderImageRepeat = exports.BORDER_IMAGE_REPEAT = void 0;
const parser_1 = require("../syntax/parser");
var BORDER_IMAGE_REPEAT;
(function (BORDER_IMAGE_REPEAT) {
    BORDER_IMAGE_REPEAT[BORDER_IMAGE_REPEAT["STRETCH"] = 0] = "STRETCH";
    BORDER_IMAGE_REPEAT[BORDER_IMAGE_REPEAT["REPEAT"] = 1] = "REPEAT";
    BORDER_IMAGE_REPEAT[BORDER_IMAGE_REPEAT["ROUND"] = 2] = "ROUND";
    BORDER_IMAGE_REPEAT[BORDER_IMAGE_REPEAT["SPACE"] = 3] = "SPACE";
})(BORDER_IMAGE_REPEAT || (exports.BORDER_IMAGE_REPEAT = BORDER_IMAGE_REPEAT = {}));
const parseRepeatValue = (value) => {
    switch (value) {
        case 'repeat':
            return BORDER_IMAGE_REPEAT.REPEAT;
        case 'round':
            return BORDER_IMAGE_REPEAT.ROUND;
        case 'space':
            return BORDER_IMAGE_REPEAT.SPACE;
        case 'stretch':
        default:
            return BORDER_IMAGE_REPEAT.STRETCH;
    }
};
exports.borderImageRepeat = {
    name: 'border-image-repeat',
    initialValue: 'stretch',
    prefix: false,
    type: 1 /* PropertyDescriptorParsingType.LIST */,
    parse: (_context, tokens) => {
        const filtered = tokens.filter(parser_1.nonWhiteSpace);
        const values = [];
        for (const token of filtered) {
            if ((0, parser_1.isIdentToken)(token)) {
                values.push(parseRepeatValue(token.value));
            }
        }
        const horizontal = values[0] ?? BORDER_IMAGE_REPEAT.STRETCH;
        const vertical = values[1] ?? horizontal;
        return { horizontal, vertical };
    }
};
