// html2canvas-pro/dist/lib/css/property-descriptors/counter-increment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.counterIncrement = exports.parseCounterValue = void 0;
const parser_1 = require("../syntax/parser");
/** Shared counter-ID parsing: returns next number or a fallback default. */
const parseCounterValue = (tokens, defaultNumber) => {
    const entries = [];
    const filtered = tokens.filter(parser_1.nonWhiteSpace);
    for (let i = 0; i < filtered.length; i++) {
        const counter = filtered[i];
        const next = filtered[i + 1];
        if (counter.type === 20 /* TokenType.IDENT_TOKEN */) {
            const num = next && (0, parser_1.isNumberToken)(next) ? next.number : defaultNumber;
            entries.push([counter.value, num]);
        }
    }
    return entries;
};
exports.parseCounterValue = parseCounterValue;
exports.counterIncrement = {
    name: 'counter-increment',
    initialValue: 'none',
    prefix: true,
    type: 1 /* PropertyDescriptorParsingType.LIST */,
    parse: (_context, tokens) => {
        if (tokens.length === 0) {
            return null;
        }
        const first = tokens[0];
        if (first.type === 20 /* TokenType.IDENT_TOKEN */ && first.value === 'none') {
            return null;
        }
        return (0, exports.parseCounterValue)(tokens, 1).map(([counter, increment]) => ({ counter, increment }));
    }
};
