// html2canvas-pro/dist/lib/css/property-descriptors/filter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.filter = void 0;
exports.filter = {
    name: 'filter',
    initialValue: 'none',
    prefix: false,
    type: 1 /* PropertyDescriptorParsingType.LIST */,
    parse: (_context, tokens) => {
        if (tokens.length === 1 && tokens[0].type === 20 /* TokenType.IDENT_TOKEN */ && tokens[0].value === 'none') {
            return null;
        }
        const parts = [];
        for (const token of tokens) {
            if (token.type === 18 /* TokenType.FUNCTION */) {
                const fn = token;
                // Reconstruct the function string from its name + values
                const renderedArgs = renderFilterArgs(fn.values);
                // Canvas API supports the same CSS filter string format
                switch (fn.name) {
                    case 'blur':
                        parts.push(`blur(${renderedArgs}px)`);
                        break;
                    case 'brightness':
                    case 'contrast':
                    case 'invert':
                    case 'opacity':
                    case 'saturate':
                    case 'sepia':
                        parts.push(`${fn.name}(${renderedArgs})`);
                        break;
                    case 'grayscale':
                        parts.push(`grayscale(${renderedArgs})`);
                        break;
                    case 'hue-rotate':
                        parts.push(`hue-rotate(${renderedArgs}deg)`);
                        break;
                    case 'drop-shadow':
                        parts.push(`drop-shadow(${renderedArgs})`);
                        break;
                    default:
                        break;
                }
            }
        }
        return parts.length > 0 ? parts.join(' ') : null;
    }
};
/**
 * Render filter function arguments back into a CSS string suitable for Canvas API.
 * Canvas 2D `ctx.filter` accepts the same CSS filter function strings.
 */
const renderFilterArgs = (values) => {
    const parts = [];
    for (const v of values) {
        if (v.type === 31 /* TokenType.WHITESPACE_TOKEN */) {
            continue;
        }
        if (v.type === 15 /* TokenType.DIMENSION_TOKEN */) {
            parts.push(`${v.number}${v.unit}`);
        }
        else if (v.type === 17 /* TokenType.NUMBER_TOKEN */) {
            parts.push(`${v.number}`);
        }
        else if (v.type === 16 /* TokenType.PERCENTAGE_TOKEN */) {
            parts.push(`${v.number}%`);
        }
        else if (v.type === 20 /* TokenType.IDENT_TOKEN */) {
            parts.push(v.value);
        }
        else if (v.type === 5 /* TokenType.HASH_TOKEN */) {
            parts.push(`#${v.value}`);
        }
    }
    return parts.join(' ');
};
