// html2canvas-pro/dist/lib/css/property-descriptors/font-variant-ligatures.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fontVariantLigatures = void 0;
exports.fontVariantLigatures = {
    name: 'font-variant-ligatures',
    initialValue: 'normal',
    prefix: false,
    type: 2 /* PropertyDescriptorParsingType.IDENT_VALUE */,
    parse: (_context, value) => {
        switch (value) {
            case 'none':
                return 1 /* FONT_VARIANT_LIGATURES.NONE */;
            case 'common-ligatures':
                return 2 /* FONT_VARIANT_LIGATURES.COMMON_LIGATURES */;
            case 'no-common-ligatures':
                return 3 /* FONT_VARIANT_LIGATURES.NO_COMMON_LIGATURES */;
            case 'discretionary-ligatures':
                return 4 /* FONT_VARIANT_LIGATURES.DISCRETIONARY_LIGATURES */;
            case 'no-discretionary-ligatures':
                return 5 /* FONT_VARIANT_LIGATURES.NO_DISCRETIONARY_LIGATURES */;
            case 'historical-ligatures':
                return 6 /* FONT_VARIANT_LIGATURES.HISTORICAL_LIGATURES */;
            case 'no-historical-ligatures':
                return 7 /* FONT_VARIANT_LIGATURES.NO_HISTORICAL_LIGATURES */;
            case 'contextual':
                return 8 /* FONT_VARIANT_LIGATURES.CONTEXTUAL */;
            case 'no-contextual':
                return 9 /* FONT_VARIANT_LIGATURES.NO_CONTEXTUAL */;
            case 'normal':
            default:
                return 0 /* FONT_VARIANT_LIGATURES.NORMAL */;
        }
    }
};
