// html2canvas-pro/dist/lib/css/property-descriptors/object-position.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.objectPosition = void 0;
const parser_1 = require("../syntax/parser");
const length_percentage_1 = require("../types/length-percentage");
exports.objectPosition = {
    name: 'object-position',
    initialValue: '50% 50%',
    prefix: false,
    type: 1 /* PropertyDescriptorParsingType.LIST */,
    parse: (_context, tokens) => {
        const result = [];
        let i = 0;
        while (i < tokens.length && result.length < 2) {
            const token = tokens[i];
            if ((0, parser_1.isIdentToken)(token)) {
                switch (token.value) {
                    case 'left':
                    case 'top':
                        result.push(length_percentage_1.ZERO_LENGTH);
                        break;
                    case 'center':
                        result.push(length_percentage_1.FIFTY_PERCENT);
                        break;
                    case 'right':
                    case 'bottom':
                        result.push(length_percentage_1.HUNDRED_PERCENT);
                        break;
                }
            }
            else if ((0, length_percentage_1.isLengthPercentage)(token)) {
                result.push(token);
            }
            i++;
        }
        // Fill to 2 values with defaults
        while (result.length < 2) {
            result.push(length_percentage_1.FIFTY_PERCENT);
        }
        return [result[0], result[1]];
    }
};
