// html2canvas-pro/dist/lib/css/property-descriptors/writing-mode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.writingMode = exports.isSidewaysWritingMode = exports.isVerticalWritingMode = void 0;
const isVerticalWritingMode = (writingMode) => writingMode !== 0 /* WRITING_MODE.HORIZONTAL_TB */;
exports.isVerticalWritingMode = isVerticalWritingMode;
const isSidewaysWritingMode = (writingMode) => writingMode === 3 /* WRITING_MODE.SIDEWAYS_RL */ || writingMode === 4 /* WRITING_MODE.SIDEWAYS_LR */;
exports.isSidewaysWritingMode = isSidewaysWritingMode;
exports.writingMode = {
    name: 'writing-mode',
    initialValue: 'horizontal-tb',
    prefix: false,
    type: 2 /* PropertyDescriptorParsingType.IDENT_VALUE */,
    parse: (_context, writingMode) => {
        switch (writingMode) {
            case 'vertical-rl':
                return 1 /* WRITING_MODE.VERTICAL_RL */;
            case 'vertical-lr':
                return 2 /* WRITING_MODE.VERTICAL_LR */;
            case 'sideways-rl':
                return 3 /* WRITING_MODE.SIDEWAYS_RL */;
            case 'sideways-lr':
                return 4 /* WRITING_MODE.SIDEWAYS_LR */;
            case 'horizontal-tb':
            default:
                return 0 /* WRITING_MODE.HORIZONTAL_TB */;
        }
    }
};
