// html2canvas-pro/dist/lib/dom/node-type-guards.js
"use strict";
/**
 * DOM Node Type Guards
 *
 * All DOM node type guards consolidated here to eliminate duplication
 * and provide a single source of truth for node type checking.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.canHavePseudoElements = exports.isCustomElement = exports.isOLElement = exports.isLIElement = exports.isSlotElement = exports.isSelectElement = exports.isTextareaElement = exports.isScriptElement = exports.isStyleElement = exports.isIFrameElement = exports.isImageElement = exports.isVideoElement = exports.isCanvasElement = exports.isBodyElement = exports.isSVGElement = exports.isHTMLElement = exports.isInputElement = exports.isHTMLElementNode = exports.isSVGElementNode = exports.isTextNode = exports.isElementNode = void 0;
// --- Generic node type guards ---
const isElementNode = (node) => node.nodeType === Node.ELEMENT_NODE;
exports.isElementNode = isElementNode;
const isTextNode = (node) => node.nodeType === Node.TEXT_NODE;
exports.isTextNode = isTextNode;
const isSVGElementNode = (element) => typeof element.className === 'object';
exports.isSVGElementNode = isSVGElementNode;
const isHTMLElementNode = (node) => (0, exports.isElementNode)(node) && typeof node.style !== 'undefined' && !(0, exports.isSVGElementNode)(node);
exports.isHTMLElementNode = isHTMLElementNode;
// --- Tag name based type guards ---
const isInputElement = (node) => node.tagName === 'INPUT';
exports.isInputElement = isInputElement;
const isHTMLElement = (node) => node.tagName === 'HTML';
exports.isHTMLElement = isHTMLElement;
const isSVGElement = (node) => node.tagName === 'svg';
exports.isSVGElement = isSVGElement;
const isBodyElement = (node) => node.tagName === 'BODY';
exports.isBodyElement = isBodyElement;
const isCanvasElement = (node) => node.tagName === 'CANVAS';
exports.isCanvasElement = isCanvasElement;
const isVideoElement = (node) => node.tagName === 'VIDEO';
exports.isVideoElement = isVideoElement;
const isImageElement = (node) => node.tagName === 'IMG';
exports.isImageElement = isImageElement;
const isIFrameElement = (node) => node.tagName === 'IFRAME';
exports.isIFrameElement = isIFrameElement;
const isStyleElement = (node) => node.tagName === 'STYLE';
exports.isStyleElement = isStyleElement;
const isScriptElement = (node) => node.tagName === 'SCRIPT';
exports.isScriptElement = isScriptElement;
const isTextareaElement = (node) => node.tagName === 'TEXTAREA';
exports.isTextareaElement = isTextareaElement;
const isSelectElement = (node) => node.tagName === 'SELECT';
exports.isSelectElement = isSelectElement;
const isSlotElement = (node) => node.tagName === 'SLOT';
exports.isSlotElement = isSlotElement;
const isLIElement = (node) => node.tagName === 'LI';
exports.isLIElement = isLIElement;
const isOLElement = (node) => node.tagName === 'OL';
exports.isOLElement = isOLElement;
const isCustomElement = (element) => !(0, exports.isSVGElementNode)(element) && element.tagName.indexOf('-') > 0;
exports.isCustomElement = isCustomElement;
const VOID_OR_REPLACED_TAGS = new Set([
    'IMG',
    'VIDEO',
    'AUDIO',
    'CANVAS',
    'IFRAME',
    'INPUT',
    'TEXTAREA',
    'SELECT',
    'BR',
    'HR',
    'META',
    'LINK',
    'BASE',
    'COL',
    'SOURCE',
    'TRACK',
    'WBR',
    'AREA',
    'PARAM',
    'EMBED',
    'OBJECT'
]);
/**
 * Check if an element can have ::before / ::after pseudo-elements.
 * Per the CSS spec, replaced elements and void elements cannot have pseudo-elements.
 * SVG elements also do not support pseudo-elements.
 */
const canHavePseudoElements = (element) => !(0, exports.isSVGElementNode)(element) && !VOID_OR_REPLACED_TAGS.has(element.tagName);
exports.canHavePseudoElements = canHavePseudoElements;
